### PR TITLE
out_forward : register config map

### PR DESCRIPTION
**Pull request by aSquare14**
Signed-off-by: Atibhi Agrawal <Atibhi.Agrawal@iiitb.org>
#1672

### CODE REVIEW

**Comment by fujimotos** [comment]
@aSquare14 I took a brief look at this patch and noticed that the config map
lacks a several existing parameters (e.g. `Username` and `Password`). I think
this needs fixing (or this changeset would make these parameters unusable).

For example, consider the following configuration:

```ini
[INPUT]
  Name Dummy
  Tag debug.log

[OUTPUT]
  Name Forward
  Host 127.0.0.1
  Port 24224
  Shared_Key foo
  Username user
  Password pass
```

and you should be able to confirm that it's not working anymore on 5e3bddf:

```
$ fluent-bit -c example.conf
...
[2020/01/27 12:33:40] [error] [config] forward: unknown configuration property 'Username'. The following properties are allowed: secured, time_as_integer, shared_key, empty_shared_key, and require_ack_response.
```

**Comment by aSquare14** [comment]
@fujimotos I am unable to reproduce the error. Could you please explain more. Thanks

**Comment by fujimotos** [comment]
@aSquare14 Thank you for the update.

I added a few comments on issues I noticed while reading your patch.
I think we can merge this patch into master once these issues are resolved,
so please have a look on them...

**Comment by aSquare14** [comment]
> I added a few comments on issues I noticed while reading your patch.
> I think we can merge this patch into master once these issues are resolved,
> so please have a look on them...

@fujimotos I can't see the new comments :/

**Comment by aSquare14** [comment]
@fujimotos Sorry for the late update!

**Comment by edsiper** [comment]
I noticed that password is not being written to the context, why ?

@fujimotos can you re-check pls ?

**Comment by fujimotos** [comment]
> I noticed that password is not being written to the context, why ?

@edsiper  I re-checked the patch and noticed that this patch never calls
the property setter function `flb_output_config_map_set()`. I assume this
is causing the problem.

@aSquare14 Can you add a call for the function to this patch? Something
like the following should make it work fine.

```diff
diff --git a/plugins/out_forward/forward.c b/plugins/out_forward/forward.c
index 094543c5..d8fafa66 100644
--- a/plugins/out_forward/forward.c
+++ b/plugins/out_forward/forward.c
@@ -721,6 +721,9 @@ static int forward_config_simple(struct flb_forward *ctx,
     }
     fc->secured = FLB_FALSE;
 
+    /* Set default values */
+    flb_output_config_map_set(ins, fc);
+
     /* Check if TLS is enabled */
 #ifdef FLB_HAVE_TLS
     if (ins->use_tls == FLB_TRUE) {
```

File: plugins/out_forward/forward.c [hunk 1]
```diff
@@ -1131,6 +1114,56 @@ static void cb_forward_flush(const void *data, size_t bytes,
...
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_STR, "username", "",
+     0, FLB_TRUE, offsetof(struct flb_forward_config,username),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_STR, "password", "",
+     0, FLB_TRUE, 0,
```
Comment (edsiper) [inline plugins/out_forward/forward.c]: if we set a default empty string in password "" as the original code states, here we are not processing that value, where is the offsetof ?

Comment (fujimotos) [reply plugins/out_forward/forward.c]: ... and the `upstream` below too. Sorry, I overlooked it.

**Comment by aSquare14** [comment]
@fujimotos Have made the changes and tested properly.

File: plugins/out_forward/forward.c [hunk 2]
```diff
@@ -1131,6 +1114,56 @@ static void cb_forward_flush(const void *data, size_t bytes,
...
+     0, FLB_TRUE, offsetof(struct flb_forward_config, empty_shared_key),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_INT, "send_options", NULL,
+     0, FLB_TRUE, offsetof(struct flb_forward_config,send_options),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_INT, "require_ack_response", NULL,
```
Comment (edsiper) [inline plugins/out_forward/forward.c]: this must be a bool type value, not integer. A bool accepts one of on, off, true or false

Comment (edsiper) [reply plugins/out_forward/forward.c]: is there any default value ?

File: plugins/out_forward/forward.c [hunk 3]
```diff
@@ -1131,6 +1114,56 @@ static void cb_forward_flush(const void *data, size_t bytes,
...
+     0, FLB_TRUE, offsetof(struct flb_forward_config, empty_shared_key),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_INT, "send_options", NULL,
+     0, FLB_TRUE, offsetof(struct flb_forward_config,send_options),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_BOOL, "require_ack_response", "FLB_FALSE",
```
Comment (edsiper) [inline plugins/out_forward/forward.c]: FLB_FALSE is an internal macro used in the code, it's not right to use it as a default value in the configmap.

Please review the config map function that cares about data type conversion and adjust it properly:

https://github.com/fluent/fluent-bit/blob/master/src/flb_config_map.c#L117

File: plugins/out_forward/forward.h [hunk 4]
```diff
@@ -45,9 +45,9 @@ struct flb_forward_config {
 
     /* config */
     flb_sds_t shared_key;     /* shared key                   */
-    flb_sds_t self_hostname;  /* hotname used in certificate  */
+    flb_sds_t self_hostname;  /* hostname used in certificate  */
     int empty_shared_key;     /* use an empty string as shared key */
-    int require_ack_response; /* Require acknowledge for "chunk" */
+    bool require_ack_response; /* Require acknowledge for "chunk" */
```
Comment (edsiper) [inline plugins/out_forward/forward.h]: we don't use ```bool``` in our source code, please set it as ```int```

File: conf/fluent-bit.conf [hunk 5]
```diff
@@ -45,4 +45,4 @@
 
 [OUTPUT]
     Name  stdout
-    Match *
+    Match *
```
Comment (edsiper) [inline conf/fluent-bit.conf]: revert this change

Comment (aSquare14) [reply conf/fluent-bit.conf]: @edsiper It is the same as master, have copied it multiple times I don't know why it is showing diff, I have removed the space also.

File: plugins/out_forward/forward.c [hunk 6]
```diff
@@ -1131,6 +1114,56 @@ static void cb_forward_flush(const void *data, size_t bytes,
...
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_INT, "empty_shared_key", NULL,
+     0, FLB_TRUE, offsetof(struct flb_forward_config, empty_shared_key),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_INT, "send_options", NULL,
+     0, FLB_TRUE, offsetof(struct flb_forward_config,send_options),
```
Comment (edsiper) [inline plugins/out_forward/forward.c]: add a space after the comma (send_options)

File: plugins/out_forward/forward.c [hunk 7]
```diff
@@ -1131,6 +1114,56 @@ static void cb_forward_flush(const void *data, size_t bytes,
...
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_BOOL, "require_ack_response", "false",
+     0, FLB_TRUE, offsetof(struct flb_forward_config,require_ack_response),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_STR, "username", "",
+     0, FLB_TRUE, offsetof(struct flb_forward_config,username),
```
Comment (edsiper) [inline plugins/out_forward/forward.c]: add missing space inside offsetof

File: plugins/out_forward/forward.c [hunk 8]
```diff
@@ -1131,6 +1114,56 @@ static void cb_forward_flush(const void *data, size_t bytes,
...
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_STR, "password", "",
+     0, FLB_TRUE, offsetof(struct flb_forward_config,password),
```
Comment (edsiper) [inline plugins/out_forward/forward.c]: add space

File: plugins/out_forward/forward.c [hunk 9]
```diff
@@ -1131,6 +1114,56 @@ static void cb_forward_flush(const void *data, size_t bytes,
...
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_STR, "upstream", NULL,
+     0, FLB_TRUE, offsetof(struct flb_forward,u),
+     NULL
```
Comment (edsiper) [inline plugins/out_forward/forward.c]: the field ```u``` on structure flb_forward is an upstream context that is populated upon initialization. When running this using an upstream configuration the service will crash:

https://github.com/fluent/fluent-bit/blob/master/plugins/out_forward/forward.c#L866

did you test this ?

File: plugins/out_forward/forward.c [hunk 10]
```diff
@@ -1138,6 +1121,56 @@ static void cb_forward_flush(const void *data, size_t bytes,
...
+     0, FLB_TRUE, offsetof(struct flb_forward_config, self_hostname),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_INT, "empty_shared_key", NULL,
+     0, FLB_TRUE, offsetof(struct flb_forward_config, empty_shared_key),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_INT, "send_options", NULL,
```
Comment (edsiper) [inline plugins/out_forward/forward.c]: this must be a BOOL

**Comment by edsiper** [comment]
@fujimotos please review the latest changes with further functionality tests.

I want to know if this is ready to go in 1.4 release or better to defer it for later.

**Comment by fujimotos** [comment]
@edsiper  Sure. I'm going to test this patch and reply tomorrow. WFM.

File: plugins/out_forward/forward.c [hunk 11]
```diff
@@ -1138,6 +1121,56 @@ static void cb_forward_flush(const void *data, size_t bytes,
...
+     0, FLB_TRUE, offsetof(struct flb_forward_config, shared_key),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_STR, "self_hostname", "localhost",
+     0, FLB_TRUE, offsetof(struct flb_forward_config, self_hostname),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_BOOL, "empty_shared_key", "true",
```
Comment (fujimotos) [inline plugins/out_forward/forward.c]: This option should be default to "false" (not "true"). Please refer to the following doc:

https://docs.fluentbit.io/manual/v/1.4-dev/pipeline/outputs/forward#secure-forward-mode-configuration-parameters

File: plugins/out_forward/forward.c [hunk 12]
```diff
@@ -1138,6 +1121,56 @@ static void cb_forward_flush(const void *data, size_t bytes,
...
+     0, FLB_TRUE, offsetof(struct flb_forward_config, time_as_integer),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_STR, "shared_key", NULL,
+     0, FLB_TRUE, offsetof(struct flb_forward_config, shared_key),
+     NULL
+    },
+    {
+     FLB_CONFIG_MAP_STR, "self_hostname", "localhost",
```
Comment (fujimotos) [inline plugins/out_forward/forward.c]: Ditto. This option should be NULL, not localhost.

File: plugins/out_forward/forward.c [hunk 13]
```diff
@@ -647,25 +645,16 @@ static int forward_config_ha(const char *upstream_file,
...
-        else {
-            fc->shared_key = NULL;
-        }
 
         tmp = flb_upstream_node_get_property("username", node);
         if (tmp) {
             fc->username = tmp;
         }
-        else {
-            fc->username = "";
```
Comment (fujimotos) [inline plugins/out_forward/forward.c]: This assignment should not removed; This will break deployments that
assume `username` defaults to empty string.

(In particular, this is because of ha-proxy setting is not being subject to flb_output_config_map_set() in the current implementation)

File: plugins/out_forward/forward.c [hunk 14]
```diff
@@ -647,25 +645,16 @@ static int forward_config_ha(const char *upstream_file,
...
-        }
 
         tmp = flb_upstream_node_get_property("password", node);
         if (tmp) {
             fc->password = tmp;
         }
-        else {
```
Comment (fujimotos) [inline plugins/out_forward/forward.c]: Ditto with "username".

File: conf/fluent-bit.conf [hunk 15]
```diff
@@ -45,4 +45,4 @@
 
 [OUTPUT]
     Name  stdout
-    Match *
```
Comment (fujimotos) [inline conf/fluent-bit.conf]: This diff is due to "\n" at the end of the file has been trimmed.

Generally not having a newline at the end is frowned upon especially
in C projects, so you would be better not to trim the character here.

**Comment by aSquare14** [comment]
@fujimotos Have done the suggested changes

File: conf/fluent-bit.conf (hunk 15 above)
Comment (fujimotos) [reply conf/fluent-bit.conf]: @aSquare14 It does not seem to be fixed. If you cannot make this
straight using your editor, try the following:

```bash
$ git checkout master conf/fluent-bit.conf
$ git add conf/fluent-bit.conf
$ git commit --amend
```

**Comment by fujimotos** [comment]
@aSquare14 `fluent-bit.conf` does not seem to be fixed by1ec97c1.
If you cannot make this straight using your editor, try the following:

```bash
$ git checkout master conf/fluent-bit.conf
$ git add conf/fluent-bit.conf
$ git commit --amend
```

**Comment by aSquare14** [comment]
@fujimotos Thanks. Done :+1:

**Comment by fujimotos** [comment]
@aSquare14 Last one thing. Sorry I was wrong about the default value of
`self_hostname`, which indeed should be `localhost` instead of NULL.
(It turned the documentation that was wrong; I'll fix that)

With that fixed, I think this PR is good to commit.

**Comment by fujimotos** [comment]
@aSquare14 I was being wrong about the default value of `self_hostname`
which indeed should be `localhost` instead of `NULL`.

I posted a patch that fixes the wrong config definition to fluent/fluent-bit-docs/pull/271.
Can you update the config map accordingly?

After that gets fixed, this patch is OK for me.

**Comment by aSquare14** [comment]
@fujimotos @edsiper Done :+1:

**Comment by fujimotos** [comment]
@aSquare14 I took your patch and merged it into master via d6c5dc5086.

There were indeed several issues in your original patch (e.g. it had a memory
issue on shutdown due to doubel-free); So I simultaneously posted a patch
1656ef79c on top of your work.

I close this issue for now. Feel free to tell me if anything is unclear.

**Comment by aSquare14** [comment]
I have commented on the commit. :) Thank you.
